presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java

```diff
@@ -320,31 +394,57 @@ public void testInnerJoinWithSpill(boolean probeHashEnabled, List<WhenSpill> whe
         }
     }
 
+    private static void checkErrors(final TaskStateMachine taskStateMachine)
+            throws Throwable
+    {
+        if (taskStateMachine.getFailureCauses().size() > 0) {
+            throw taskStateMachine.getFailureCauses().peek();
```

Comment:
For the spill failure case, it's perfect. But if this triggers in non-failure-expecting test, then the fact you re-throw exception from another thread may be misleading. The stacktrace produced won't match the actual stack being traversed by the exception making source of exception un-traceable. 

Also, `taskStateMachine.getFailureCauses()` seems concurrent collection, so it's conceivable that `.peek()` returns `null`, thus consider

```
throw new RuntimeException(requireNotNull(taskStateMachine.getFailureCauses().peek()));
```